### DISCUSSION TITLE
Android VPN Client

### DISCUSSION THREAD

**Discussion by Smart781** [fyne-io/fyne#5109] [Q&A]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same feature, before opening a new one.
- [X] This issue only relates to a single feature. I will open new issues for any other features.

### Is your feature request related to a problem?

Hello! I am writing an Outline VPN client for Android in the Go language and use outline-go-tun2SOCKS (https://github.com/Jigsaw-Code/outline-apps/tree/master/client/go). In order to connect with the server on Android, I need a TUN file descriptor (https://github.com/Jigsaw-Code/outline-apps/blob/master/client/go/outline/tun2socks/tunnel_android.go#L43) is created on Android using VpnService (https://developer.android.com/reference/android/net/VpnService). But it cannot be launched from Go (https://pkg.go.dev/github.com/eycorsican/go-tun2socks-android), so I try to use Gomobile Bind (https://pkg.go.dev/golang.org/x/mobile/cmd/gobind) to create a method working on Go and calling vpnservice, which is made 
on Kotlin. However, during the compilation of Gomobile Bind, errors occur. Could you please, help me with that? Any help would be much appreciated!!

### Describe the solution you'd like to see.

The function that runs VpnService on Go

**Comment by andydotxyz** [comment]
We do not support "gomobile bind" or any usage of the gomobile dependency.

The Android functionality is provided within Fyne and you should use `RunNative` to access OS level APIs.

**Comment by Smart781** [comment]
> We do not support "gomobile bind" or any usage of the gomobile dependency.
> 
> The Android functionality is provided within Fyne and you should use `RunNative` to access OS level APIs.

Thank you so much! If that’s not too much to ask, could you please share a little snippet of how the code could looks like to start a VPNService using RunNative?

**Comment by andydotxyz** [comment]
This document should help https://fynelabs.com/2024/03/01/running-native-android-code-in-a-fyne-app/

**Comment by Smart781** [comment]
> This document should help https://fynelabs.com/2024/03/01/running-native-android-code-in-a-fyne-app/

Thanks for the document! I really understand better how to do this, but the example works with a static field. I tried to do it in a similar way
(https://gist.github.com/Smart781/1e6dcf3b86080e517dfa96979363da45) but got an error (https://gist.github.com/Smart781/932b2e9bcf8fa13ff129b4917c288982)
From what I’ve found (https://www.demo2s.com/android/android-vpnservice-tutorial-with-examples.html) - it seems like I need to do AndroidManifest.xml with permissions and specify in VPN Service as a separate service; it seems like it’s something specific to fyne, as it’s an abstraction layer on top of Java - could you please, advice how to better achieve that?”

**Comment by andydotxyz** [comment]
From the following on tickets this was clearly intended as a discussion, converting.

**Comment by andydotxyz** [comment]
Accessing the native code is not Fyne specific, and there have been examples posted.
If you want a custom android manifest, just make an AndroidManifest.xml in the root of your project and it will be included.
